### DISCUSSION TITLE
Keep same Footer across different pages/routes

### DISCUSSION THREAD

**Discussion by LarsEggimann** [zauberzeug/nicegui#3379] [Q&A] [answered]
### Question

Hi there!

I am using nicegui to create a userinterface for a few lab instruments (electro meter, stepper motor usw...) which looks currently like this:
![grafik](https://github.com/user-attachments/assets/af0e83f7-4726-4115-aa60-c3e46b8b3d44)

I am using footer, header and side drawers according to [the example ](https://nicegui.io/documentation/page_layout#page_layout) on the nicegui docs. 

I have a main file where i define the different pages like this:
```python
@ui.page("/chopper_wheel")
def chopper_wheel_page():
    create_layout()
    device_manager.chopper_wheel_ui()  # type: ignore


@ui.page("/ion_chamber")
def ion_chamber_page():
    create_layout()
    device_manager.ion_chamber_ui()  # type: ignore
```

Within the `create_layout()` function I have the header, footer usw which gets loaded when i switch the page.
```python
def create_layout():

   # .... header and drawers usw

    with ui.footer().classes("flex justify-between "):
        ui.label("FLASH control panel")
        times_clicked = 0
        def mysterious_button():
            a.update()
            nonlocal times_clicked
            times_clicked += 1
            if times_clicked > 2:
                a.seek(103)
                a.play()
                a.props("controls")
                ui.notify("You found an easter egg! 🥚")
        with ui.row().classes("items-center"):
            ui.label("Made with")
            a = ui.audio(
                src=str(working_directory / "static" / "spinme.mp3"), controls=False
            )
            ui.button(" ❤️ ", on_click=mysterious_button, color=None).tooltip(
                "Whats this?"
            )
            ui.label("by ...")
            with ui.link(
                target="mailto:mymail?subject=FLASH%20Control%20Software%20-%20Contact%20me%20Button&body=Hello%20there%20you%20kind%20creator%20of%20this%20magnificent%20app!%0A%0AI%20have%20found%20the%20following%20bug%3A",
                new_tab=True,
            ).tooltip("Contact Me!"):
                ui.icon("mail")
```

But this means that the footer gets reloaded everytime I switch to another page. Now my question is how to have a static footer, header and side drawers such that for example the auido file that is played from the footer does not stop when I switch pages?
(this will also be useful for other things and not just easter eggs, I just came across this problem while playing around with this)

Thanks in advance for all the help!

Cheers,
Lars

**Comment by python-and-novella** [comment, marked as answer]
Check this:
https://github.com/zauberzeug/nicegui/tree/main/examples/single_page_app/

**Reply by LarsEggimann** [reply]
thank you very much this seems to be exactly what I need. :D
Is there a reason why this router decorator is not in package? Does it have performance issues or something like that?

Thanks for your help!

**Reply by rodja** [reply]
We are working on an integrated API in https://github.com/zauberzeug/nicegui/pull/2811.

**Reply by falkoschindler** [reply]
@LarsEggimann One of the main reasons why we didn't include it in the library was a naming issue. There is already a router with a different meaning in FastAPI, which might have caused confusion. So we kept it in the repository as proof of concept.
Currently there is PR #2811 in the making, which will introduce a proper API for single page applications. Until it is finished, you can use the above-mentioned example as a workaround.

**Reply by LarsEggimann** [reply]
thank you very much for these answers! 

I love using nicegui, it is awesome. Keep up the good work.
